### DISCUSSION TITLE
Does providing testValue mean you don't need to explicitly override dependency in tests?

### DISCUSSION THREAD

**Discussion by theMuslimDev** [pointfreeco/swift-dependencies#93] [Q&A] [answered]
I was going through the documentation and came across this:

```swift
withDependencies {
  $0.context = .test
} operation: {
  // Dependencies accessed here default to their "test" value
}
```

Does this mean if I provide a `testValue` when registering my dependency. I don't need to override dependency inside the `withDepedency`. If the answer is yes. I'm having some issues where I'm getting an error saying there is no test implementation for my dependency when running a test. I'm doing something like this.

```swift
class SomeClass: FirstProtocol { /* ... */ }

enum FirstDependencyKey: DependencyKey {
  static let liveValue: any FirstProtocol = SomeClass()
}

extension DependencyValues {
 var apiClientOne: FirstProtocol {
    get { self[FirstDependencyKey.self] }
    set { self[FirstDependencyKey.self] = newValue }
  }
}

class SomeViewModel: ObservableObject {
    @Published var isAwesome: Bool = false
    @Dependency(\.apiClientOne) var apiClientOne
    
    func onButtonTapped() {
        apiClientOne.fetchUser {
            isAwesome = true
            
        }
    }
}


// Inside unit test target
extension FirstDependencyKey {
    static let testValue: any FirstProtocol = /* ... */
}

func test_someAction_in_someClass() {
    // Given
    let sut = withDependencies {
        $0.context = .test
    } operation: {
        SomeViewModel()
    }
    
    // When
    sut.onButtonTapped()
    
    // Then
    XCTAssertTrue(sut.isAwesome)
}

```

In example like above. I keep getting error like:

```swift 
test_someAction_in_someClass(): @Dependency(\. apiClientOne) has no test implementation, but was accessed from a test context:
```

I'm not sure what I'm doing incorrect.

**Comment by tgrapperon** [comment]
Hey @Muhammed9991!
You're seeing this message because you only provided a `liveValue` implementation. One way to fix this is to provide an explicit value for the dependency in the `withDependencies` block.
You can also provide a `testValue` that will automatically be used when testing. The convention is to make this value "unimplemented" by default using `XCTestDynamicOverlay` (check built-in dependencies for some examples), but there are a few cases where you may want to provide a non-failing implementation by default.
You usually don't need to touch `.context`. When testing, `testValue`s are selected by default. If you implemented them as "unimplemented", you'll need to override them with explicit values in `withDependencies`'s closure.

**Reply by theMuslimDev** [reply]
In the extension I did provide a testValue. But it's an extension inside the test target. I've had a look at the recommended default using unimplemented. But in this case I wanted to provide some mock data to the `testValue`

**Reply by tgrapperon** [reply, marked as answer]
This is a gotcha: you need to define all the requirements in the same module in your case. You can conform to `TestDependencyKey` and then later to `DependencyKey` in a parent module, but once you conform to `DependencyKey`, all the values are set and the ones that you declared in the tests are not considered as fulfillments of the protocol.

**Reply by theMuslimDev** [reply]
Ohh. I might have missed that in the documentation. 

For context, I am using sourcery to generate mocks. Which is part of test target only. That's why the extension was in the test target. I'll play around with it and see if that resolves this issue. 

Thanks for the the prompt reply @tgrapperon 🙌🏽

**Reply by tgrapperon** [reply]
Not having at least warnings at build time is a deficiency of Swift. I think that they plan to improve the situation for Swift 6 though.

**Reply by yan-zaitsev-hs** [reply]
Hi @Muhammed9991 did you find any solution to your request? We stuck on the same. Want to have testValue defined and mocks in test target only

**Reply by theMuslimDev** [reply]
You have to do something like this:

``` swift 
// Module: AnalyticsClient
struct AnalyticsClient: TestDependencyKey {
  // ...

  static let testValue = Self(/* ... */)
}

```
And then in the implementation module you can extend the dependency to further conform to the [DependencyKey](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/dependencykey) protocol and provide a live implementation:

```swift
// Module: LiveAnalyticsClient
extension AnalyticsClient: DependencyKey {
  static let liveValue = Self(/* ... */)
}
```

More information in the documentation [here](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/livepreviewtest#Test-value) under subheading `Separating interface and implementation`
